types/parser_driver/value_expr.go

```diff
@@ -96,7 +96,8 @@ func (n *ValueExpr) Restore(ctx *format.RestoreCtx) error {
 	case types.KindFloat64:
 		ctx.WritePlain(strconv.FormatFloat(n.GetFloat64(), 'e', -1, 64))
 	case types.KindString:
-		if n.Type.Charset != "" {
+		if n.Type.Charset != "" &&
+			(!ctx.Flags.HasStringWithoutDefaultCharset() || n.Type.Charset != mysql.DefaultCharset) {
```

Comment:
How is this change related to this PR?

Reply:
This is for the content after the restore to be compatible with the text before this PR, that is, the default charset is not printed.